File: src/UE4_Assets/exports/FortCosmeticVariant.java
/**
 * 
 */
package UE4_Assets.exports;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import UE4.deserialize.NewableWithUObject;
import UE4_Assets.FText;
import UE4_Assets.ImportMap;
import UE4_Localization.Locres;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * @author FunGames
 *
 */
@Data
@NoArgsConstructor
public class FortCosmeticVariant implements NewableWithUObject {
	private FText variantChannelName;
	private String variantChannelTag;
	private List<CosmeticVariantContainer> variants;
	private UObject baseObject;

	@Override
	public void init(UObject uObject, Gson gson, ImportMap importMap) {
		JsonObject ob = gson.toJsonTree(uObject).getAsJsonObject();
		Optional.ofNullable(ob.get("VariantChannelTag")).ifPresent(e -> {
			variantChannelTag = e.getAsJsonObject().get("TagName").getAsString();
		});
		Optional.ofNullable(ob.get("VariantChannelName")).ifPresent(e -> {
			JsonObject textProps = e.getAsJsonObject();
			variantChannelName = new FText(textProps.get("namespace").getAsString(),
					textProps.get("key").getAsString(), textProps.get("text").getAsString());
		});

		String searchTag = null;
		switch (uObject.getExportType()) {
		case "FortCosmeticCharacterPartVariant":
			searchTag = "PartOptions";
			break;
		case "FortCosmeticMaterialVariant":
			searchTag = "MaterialOptions";
			break;
		case "FortCosmeticParticleVariant":
			searchTag = "ParticleOptions";
			break;
		default:
			searchTag = uObject.getExportType().replace("FortCosmetic", "").replace("Variant", "");
			break;
		}
		
		Optional.ofNullable(ob.get(searchTag)).ifPresent(e -> {
			JsonArray vars = e.getAsJsonArray();
			variants = new ArrayList<>();
			vars.forEach(e1 -> {
				JsonObject o = e1.getAsJsonObject();
				CosmeticVariantContainer v = new CosmeticVariantContainer();
				v.setBHideIfNotOwned(o.get("bHideIfNotOwned").getAsBoolean());
				v.setBIsDefault(o.get("bIsDefault").getAsBoolean());
				v.setBStartUnlocked(o.get("bStartUnlocked").getAsBoolean());
				v.setCustomizationVariantTag(o.get("CustomizationVariantTag").getAsJsonObject().get("TagName").getAsString());
				v.setPreviewImage(o.get("PreviewImage").getAsJsonObject().get("asset_path").getAsString());
				JsonObject nameProps = o.get("VariantName").getAsJsonObject();
				v.setVariantName(new FText(nameProps.get("namespace").getAsString(),
						nameProps.get("key").getAsString(), nameProps.get("text").getAsString()));
				variants.add(v);
			});
		});
	}
	/**
	 * @param locres
	 * @return
	 */
	public FortCosmeticVariant forLanguage(Optional<Locres> locres) {
		if (variantChannelName != null)
			this.variantChannelName = this.variantChannelName.cloneForLocres(locres);
		variants.forEach(varContainer -> {
			varContainer.setVariantName(varContainer.getVariantName().cloneForLocres(locres));
		});
		return this;
	}

}

File: src/main/kotlin/me/fungames/jfortniteparse/ue4/assets/exports/fort/FortCosmeticVariant.kt
package me.fungames.jfortniteparse.ue4.assets.exports.fort

import me.fungames.jfortniteparse.ue4.assets.exports.UExport
import me.fungames.jfortniteparse.ue4.assets.exports.UObject
import me.fungames.jfortniteparse.ue4.assets.objects.*
import me.fungames.jfortniteparse.ue4.assets.reader.FAssetArchive
import me.fungames.jfortniteparse.ue4.assets.util.FName
import me.fungames.jfortniteparse.ue4.assets.writer.FAssetArchiveWriter
import me.fungames.jfortniteparse.ue4.locres.Locres
import java.awt.image.BufferedImage


data class CosmeticVariant(var startUnlocked : Boolean, var isDefault : Boolean,
                           var hideIfNotOwned : Boolean, var customizationVariantTag : FName,
                           var variantName : FText?, var previewImage : FSoftObjectPath?) {
    var previewIcon : BufferedImage? = null
}

@ExperimentalUnsignedTypes
class FortCosmeticVariant : UExport {
    override var baseObject: UObject
    var variantChannelName : FText? = null
    var variantChannelTag : FName? = null
    val variants : MutableList<CosmeticVariant>

    constructor() : super("FortCosmeticVariant") {
        baseObject = UObject(
            mutableListOf(),
            null,
            "ItemDefinition"
        )
        variants = mutableListOf()
    }

    constructor(Ar: FAssetArchive, exportObject: FObjectExport) : super(exportObject) {
        super.init(Ar)
        baseObject = UObject(Ar, exportObject)
        variantChannelName = baseObject.getOrNull("VariantChannelName")
        variantChannelTag = baseObject.getOrNull<FStructFallback>("VariantChannelTag")?.getOrNull<FName>("TagName")
        val searchTag = when(exportObject.classIndex.importName) {
            "FortCosmeticCharacterPartVariant" -> "PartOptions"
            "FortCosmeticMaterialVariant" -> "MaterialOptions"
            "FortCosmeticParticleVariant" -> "ParticleOptions"
            else -> exportObject.classIndex.importName
                .replace("FortCosmetic", "")
                .replace("Variant", "")
        }
        variants = mutableListOf()
        val options = baseObject.getOrNull<UScriptArray>(searchTag) ?: return
        options.data.filterIsInstance<FStructFallback>().forEach {
            val startUnlocked = it.getOrDefault("bStartUnlocked", false)
            val isDefault = it.getOrDefault("bIsDefault", false)
            val hideIfNotOwned = it.getOrDefault("bHideIfNotOwned", false)
            val customizationVariantTag = it.get<FStructFallback>("CustomizationVariantTag").get<FName>("TagName")
            val variantName = it.getOrNull<FText>("VariantName")
            val previewImage = it.getOrNull<FSoftObjectPath>("PreviewImage")
            variants.add(CosmeticVariant(startUnlocked, isDefault, hideIfNotOwned, customizationVariantTag, variantName, previewImage))
        }
        super.complete(Ar)
    }

    override fun serialize(Ar: FAssetArchiveWriter) {
        super.initWrite(Ar)
        baseObject.serialize(Ar)
        super.completeWrite(Ar)
    }

    override fun applyLocres(locres : Locres?) {
        variantChannelName?.applyLocres(locres)
        variants.forEach {
            it.variantName?.applyLocres(locres)
        }
    }
}